README.md
![npm](https://img.shields.io/npm/dt/cordova-plugin-biometric-auth) ![npm](https://img.shields.io/npm/v/cordova-plugin-biometric-auth) ![GitHub package.json version](https://img.shields.io/github/package-json/v/andreszs/cordova-plugin-biometric-auth?color=FF6D00&label=master&logo=github) ![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/andreszs/cordova-plugin-biometric-auth) ![GitHub top language](https://img.shields.io/github/languages/top/andreszs/cordova-plugin-biometric-auth) ![GitHub](https://img.shields.io/github/license/andreszs/cordova-plugin-biometric-auth) ![GitHub last commit](https://img.shields.io/github/last-commit/andreszs/cordova-plugin-biometric-auth)

# cordova-plugin-biometric-auth

Biometric authentication with optional KeyguardManager API for Cordova.

# Platforms

- Android 5+
- Browser (filler)

# Features

- AndroidX ready
- Authenticate with [BiometricManager](https://developer.android.com/reference/androidx/biometric/BiometricManager) (fingerprint, iris, face, device credentials) since API 23
- Authenticate with [KeyguardManager](https://developer.android.com/reference/android/app/KeyguardManager) (pin, pattern, password, biometric if enrolled) since API 21
- Auto fallback to KeyguardManager when no biometric enrolled or supported
- Supports all authentication modes (WEAK, STRONG, DEVICE CREDENTIALS)
- Supports [API level](https://apilevels.com/ "API level") 21 and over

# Installation

## Install latest version from NPM

```bash
  cordova plugin add cordova-plugin-biometric-auth
```

# Methods

## isAvailable

Checks if the user can authenticate with either biometrics, fallback PIN, pattern or password. Biometric requires at least one biometric sensor to be present, enrolled, and available on the device.

```javascript
cordova.plugins.BiometricAuth.isAvailable(successCallback, errorCallback, [optionalParams])
```

| **optionalParams** | |
| --- | --- |
| authenticators | **int**: An optional bit field representing the types of [Authenticators](https://developer.android.com/reference/androidx/biometric/BiometricManager.Authenticators) that may be used for authentication on Android. Omit or use `0` to check for either biometrics or device credentials. Use `1`  to check for KeyguardManager authentication. |

### Android quirks

Not all combinations of authenticator types are supported prior to Android 11 (API 30). Specifically, `DEVICE_CREDENTIAL` alone is unsupported prior to API 30, and `BIOMETRIC_STRONG | DEVICE_CREDENTIAL` is unsupported on API 28-29.

### Browser quirks

This filler platform always returns **BIOMETRIC_SUCCESS** and does not check nor use a real biometric device.

### successCallback return values

- **BIOMETRIC_SUCCESS**: The user can authenticate with the requested method(s).
- **KEYGUARD_MANAGER**: Returned on API 21-22, or when biometric is not enrolled and `authenticators` value passed is `0` or `1`: The user can authenticate with KeyuardManager methods.

### errorCallback return values

- **BIOMETRIC_ERROR_HW_UNAVAILABLE**
- **BIOMETRIC_ERROR_NONE_ENROLLED**
- **BIOMETRIC_ERROR_NO_HARDWARE**
- **BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED**
- **BIOMETRIC_ERROR_UNSUPPORTED**
- **BIOMETRIC_STATUS_UNKNOWN**

### Example 1

Check for any biometric enrolled, PIN, pattern or password availability.

```javascript
var onSuccess = function (strSuccess) {
	console.log(strSuccess);
};
var onError = function (strError) {
	console.warn(strError);
};
cordova.plugins.BiometricAuth.isAvailable(onSuccess, onError);
```

### Example 2

Check for any biometric (e.g. fingerprint, iris, or face) on the device that meets or exceeds the requirements for Class 2. Requires at least API 23 (Android 6).

```javascript
var Authenticators = {
	KEYGUARD_MANAGER: 1,
	BIOMETRIC_STRONG: 15,
	BIOMETRIC_WEAK: 255,
	DEVICE_CREDENTIAL: 32768
};
var onSuccess = function (strSuccess) {
	console.log(strSuccess);
};
var onError = function (strError) {
	console.warn(strError);
};
var optionalParams = {
	authenticators = Authenticators.BIOMETRIC_WEAK;
};
cordova.plugins.BiometricAuth.isAvailable(onSuccess, onError, optionalParams);
```

## authenticate

Shows the biometric prompt or the fallback device credential dialog for authentication.

```javascript
cordova.plugins.BiometricAuth.authenticate(successCallback, errorCallback, [optionalParams])
```

| optionalParams | |
| --- | --- |
| title | **string**: The title to be displayed on the prompt. Defaults to *Enter unlock credentials*. |
| subtitle | **string**: The subtitle to be displayed on the prompt. |
| disableBackup | **boolean**: Removes the backup option from the prompt. Defaults to `false`. |
| **Android-specific** | |
| authenticators | **int:** A bit field representing all valid [authenticator](https://developer.android.com/reference/androidx/biometric/BiometricManager.Authenticators) types that may be invoked by the prompt. Use `0` to allow either biometrics or device credentials. Use `1`  to invoke KeyguardManager PIN, pattern, password or biometric if enrolled authentication. |
| negativeButtonText | **string**: Sets the text for the cancel button on the prompt. Required whenever fallback is disabled. |

### Android quirks

Not all combinations of authenticator types are supported prior to Android 11 (API 30). Specifically, `DEVICE_CREDENTIAL` alone is unsupported prior to API 30, and `BIOMETRIC_STRONG | DEVICE_CREDENTIAL` is unsupported on API 28-29.

### Browser quirks

Browser platforms shows a dialog to manually select either of these results and does not perform any actual biometric check:

 - **AUTHENTICATION_FAILED**
 - **BIOMETRIC_DISMISSED**
 - **BIOMETRIC_SUCCESS**

### successCallback return values

- **AUTHENTICATION_SUCCEEDED**

### errorCallback return values

- **AUTHENTICATION_FAILED**
- Please test demo app provided for other values.

### Example

Prompt the user for biometric, PIN, pattern or password credentials.

```javascript
var onSuccess = function (strSuccess) {
	console.log(strSuccess);
};
var onError = function (strError) {
	console.warn(strError);
};
var optionalParams = {
	title = "Confirm operation",
	subtitle = "Verify with biometrics to continue",
};
cordova.plugins.BiometricAuth.authenticate(onSuccess, onError, optionalParams);
```

### Remarks

- Do not use `BIOMETRIC_STRONG` without checking its availability with **isAvailable** first.
- Do not use `DEVICE_CREDENTIALS` alone prior to API 30.
- Do not use `BIOMETRIC_STRONG + DEVICE_CREDENTIAL` on API 28-29.
- To force usage of KeyguardManager instead of BiometricManager, set `1` to the **authenticators** param.
- Using an **authenticators** value other than `0` or `1` will discard the **disableBackup** option.
- Always provide a **negativeButtonText** when  using **disableBackup** or not using `DEVICE_CREDENTIAL` authenticator.
- Android 5 will use the KeyguardManager PIN, pattern or password regardless of any options.

# Plugin demo app

- [Compiled debug APK and reference](https://www.andreszsogon.com/cordova-biometric-auth-plugin-demo/ "Compiled debug APK and reference")
- [Source code for www folder](https://github.com/andreszs/cordova-plugin-demos "Source code for www folder")

<img src="https://github.com/andreszs/cordova-plugin-demos/blob/main/com.andreszs.biometric.auth.demo/screenshots/android/biometric-auth.png?raw=true" width="200" /> <img src="https://github.com/andreszs/cordova-plugin-demos/blob/main/com.andreszs.biometric.auth.demo/screenshots/android/biometric-auth-authenticators.png?raw=true" width="200" /> <img src="https://github.com/andreszs/cordova-plugin-demos/blob/main/com.andreszs.biometric.auth.demo/screenshots/android/biometric-auth-strong.png?raw=true" width="200" /> <img src="https://github.com/andreszs/cordova-plugin-demos/blob/main/com.andreszs.biometric.auth.demo/screenshots/android/biometric-auth-credentials.png?raw=true" width="200" />

# Contributing

Please report any issue with this plugin in GitHub by providing detailed context and sample code.
PRs to improve and add new features or platforms are always welcome.

- **PR to add iOS platform is welcome**

RELEASENOTES.md
# Release Notes

### 1.0.1 (Jun 13, 2022)

* Updated README and created this RELEASENOTES file.

### 1.0.0 (Jun 11, 2022)

* Initial release with Android and browser as filler platform.


src/browser/BiometricAuthProxy.js
var BiometricAuth = function () {};

BiometricAuth.prototype.AVAILABLE = "AVAILABLE";
BiometricAuth.prototype.AUTHENTICATION_FAILED = "AUTHENTICATION_FAILED";
BiometricAuth.prototype.BIOMETRIC_SUCCESS = "BIOMETRIC_SUCCESS";
BiometricAuth.prototype.BIOMETRIC_DISMISSED = "BIOMETRIC_DISMISSED";

/* Browser features */
BiometricAuth.prototype.HIGHEST_POSSIBLE_Z_INDEX = 2147483647;
BiometricAuth.prototype.FINGERPRINT_BG_DIV = 'cordova_fingerprint_plugin_bg';
BiometricAuth.prototype.FINGERPRINT_FG_DIV = 'cordova_fingerprint_plugin_fg';

/* Browser buttons */
BiometricAuth.prototype.ICON_SUCCESS_SRC = 'fingerprint-success.png';
BiometricAuth.prototype.ICON_SUCCESS_ID = 'cordova_plugin_fingerprint_icon_success';
BiometricAuth.prototype.ICON_DISMISS_SRC = 'fingerprint-chevron-left.png';
BiometricAuth.prototype.ICON_DISMISS_ID = 'cordova_plugin_fingerprint_icon_dismiss';
BiometricAuth.prototype.ICON_FAILED_SRC = 'fingerprint-close.png';
BiometricAuth.prototype.ICON_FAILED_ID = 'cordova_plugin_fingerprint_icon_failed';
BiometricAuth.prototype.ICON_SIZE = '';

BiometricAuth.prototype.authenticate = function (successCallback, errorCallback, args) {
	/* translucid background div */
	var dialogBg = document.createElement('div');
	dialogBg.id = BiometricAuth.prototype.FINGERPRINT_BG_DIV;
	dialogBg.style.backgroundColor = '#00000088';
	dialogBg.style.position = 'fixed';
	dialogBg.style.top = '0';
	dialogBg.style.bottom = '0';
	dialogBg.style.left = '0';
	dialogBg.style.right = '0';
	dialogBg.style.zIndex = (BiometricAuth.prototype.HIGHEST_POSSIBLE_Z_INDEX - 1);
	document.body.appendChild(dialogBg);

	/* foreground text div */
	var dialogFg = document.createElement('div');
	dialogFg.id = BiometricAuth.prototype.FINGERPRINT_FG_DIV;
	dialogFg.setAttribute('style', 'transform: translateY(-50%);');
	dialogFg.style.backgroundColor = '#424242';
	dialogFg.style.boxShadow = 'rgba(0,0,0,0.5) 0px 4px 24px';
	dialogFg.style.borderRadius = '3px';
	dialogFg.style.marginLeft = 'auto';
	dialogFg.style.marginRight = 'auto';
	dialogFg.style.padding = '1.5em';
	dialogFg.style.position = 'relative';
	dialogFg.style.top = '50%';
	dialogFg.style.textAlign = 'center';
	dialogFg.style.width = '70%';
	dialogFg.style.zIndex = BiometricAuth.prototype.HIGHEST_POSSIBLE_Z_INDEX;

	/* title */
	var title = '';
	if (args && typeof (args[0]) === 'object' && typeof (args[0].title) === 'string' && args[0].title.length > 0) {
		title = args[0].title;
	} else {
		title = 'BiometricAuth Sign On';
	}
	var divTitle = document.createElement('div');
	divTitle.appendChild(document.createTextNode(title));
	divTitle.style.color = '#FFFFFF';
	divTitle.style.fontSize = '20px';
	divTitle.style.fontWeight = '600';
	divTitle.style.lineHeight = '2em';
	divTitle.style.textAlign = 'center';
	dialogFg.appendChild(divTitle);

	/* optional subtitle */
	if (args && typeof (args[0]) === 'object' && typeof (args[0].subtitle) === 'string' && args[0].subtitle.length > 0) {
		var divSubtitle = document.createElement('div');
		divSubtitle.appendChild(document.createTextNode(args[0].subtitle));
		divSubtitle.style.color = '#BDBDBD';
		divSubtitle.style.fontSize = '14px';
		divSubtitle.style.lineHeight = '2em';
		divSubtitle.style.textAlign = 'center';
		dialogFg.appendChild(divSubtitle);
	}

	/* optional description */
	if (args && typeof (args[0]) === 'object' && typeof (args[0].description) === 'string' && args[0].description.length > 0) {
		var divDescription = document.createElement('div');
		divDescription.appendChild(document.createTextNode(args[0].description));
		divDescription.style.color = '#BDBDBD';
		divDescription.style.fontSize = '14px';
		divDescription.style.lineHeight = '2em';
		divDescription.style.textAlign = 'center';
		dialogFg.appendChild(divDescription);
	}

	/* BIOMETRIC_DISMISSED */
	var imgIconDismissed = document.createElement('img');
	imgIconDismissed.id = BiometricAuth.prototype.ICON_DISMISS_ID;
	imgIconDismissed.src = BiometricAuth.prototype.ICON_DISMISS_SRC;
	imgIconDismissed.style.backgroundColor = '#FFFFFF10';
	imgIconDismissed.style.borderColor = '#FFFFFF30';
	imgIconDismissed.style.borderRadius = '50%';
	imgIconDismissed.style.borderStyle = 'dashed';
	imgIconDismissed.style.borderWidth = '2px';
	imgIconDismissed.style.cursor = 'pointer';
	imgIconDismissed.style.display = 'inline-block';
	imgIconDismissed.style.margin = '8px';
	imgIconDismissed.style.marginTop = '24px';
	imgIconDismissed.style.padding = '8px';
	imgIconDismissed.style.minWidth = '32px';
	imgIconDismissed.style.maxWidth = '96px';
	imgIconDismissed.style.width = '20%';
	imgIconDismissed.title = BiometricAuth.prototype.BIOMETRIC_DISMISSED;
	imgIconDismissed.addEventListener('mouseover', function () {
		imgIconDismissed.style.backgroundColor = '#FFFFFF20';
	});
	imgIconDismissed.addEventListener('mouseout', function () {
		imgIconDismissed.style.backgroundColor = '#FFFFFF10';
	});
	imgIconDismissed.addEventListener('click', function () {
		document.getElementById(BiometricAuth.prototype.FINGERPRINT_BG_DIV).remove();
		errorCallback(BiometricAuth.prototype.BIOMETRIC_DISMISSED);
	});
	dialogFg.appendChild(imgIconDismissed);

	/* BIOMETRIC_SUCCESS */
	var imgIconSuccess = document.createElement('img');
	imgIconSuccess.id = BiometricAuth.prototype.ICON_SUCCESS_ID;
	imgIconSuccess.src = BiometricAuth.prototype.ICON_SUCCESS_SRC;
	imgIconSuccess.style.backgroundColor = '#FFFFFF10';
	imgIconSuccess.style.borderColor = '#FFFFFF30';
	imgIconSuccess.style.borderRadius = '50%';
	imgIconSuccess.style.borderStyle = 'dashed';
	imgIconSuccess.style.borderWidth = '2px';
	imgIconSuccess.style.cursor = 'pointer';
	imgIconSuccess.style.display = 'inline-block';
	imgIconSuccess.style.margin = '8px';
	imgIconSuccess.style.marginTop = '24px';
	imgIconSuccess.style.padding = '8px';
	imgIconSuccess.style.minWidth = '32px';
	imgIconSuccess.style.maxWidth = '96px';
	imgIconSuccess.style.width = '20%';
	imgIconSuccess.title = BiometricAuth.prototype.STR_BIOMETRIC_SUCCESS;
	imgIconSuccess.addEventListener('mouseover', function () {
		imgIconSuccess.style.backgroundColor = '#FFFFFF20';
	});
	imgIconSuccess.addEventListener('mouseout', function () {
		imgIconSuccess.style.backgroundColor = '#FFFFFF10';
	});
	imgIconSuccess.addEventListener('click', function () {
		document.getElementById(BiometricAuth.prototype.FINGERPRINT_BG_DIV).remove();
		successCallback(BiometricAuth.prototype.BIOMETRIC_SUCCESS);
	});
	dialogFg.appendChild(imgIconSuccess);

	/* BIOMETRIC_AUTHENTICATION_FAILED */
	var imgIconFailed = document.createElement('img');
	imgIconFailed.id = BiometricAuth.prototype.ICON_FAILED_ID;
	imgIconFailed.src = BiometricAuth.prototype.ICON_FAILED_SRC;
	imgIconFailed.style.backgroundColor = '#FFFFFF10';
	imgIconFailed.style.borderColor = '#FFFFFF30';
	imgIconFailed.style.borderRadius = '50%';
	imgIconFailed.style.borderStyle = 'dashed';
	imgIconFailed.style.borderWidth = '2px';
	imgIconFailed.style.cursor = 'pointer';
	imgIconFailed.style.display = 'inline-block';
	imgIconFailed.style.margin = '8px';
	imgIconFailed.style.marginTop = '24px';
	imgIconFailed.style.padding = '8px';
	imgIconFailed.style.minWidth = '32px';
	imgIconFailed.style.maxWidth = '96px';
	imgIconFailed.style.width = '20%';
	imgIconFailed.title = BiometricAuth.prototype.AUTHENTICATION_FAILED;
	imgIconFailed.addEventListener('mouseover', function () {
		imgIconFailed.style.backgroundColor = '#FFFFFF20';
	});
	imgIconFailed.addEventListener('mouseout', function () {
		imgIconFailed.style.backgroundColor = '#FFFFFF10';
	});
	imgIconFailed.addEventListener('click', function () {
		document.getElementById(BiometricAuth.prototype.FINGERPRINT_BG_DIV).remove();
		errorCallback(BiometricAuth.prototype.AUTHENTICATION_FAILED);
	});
	dialogFg.appendChild(imgIconFailed);

	document.getElementById(BiometricAuth.prototype.FINGERPRINT_BG_DIV).appendChild(dialogFg);
};

BiometricAuth.prototype.isAvailable = function (successCallback, errorCallback, args) {
	successCallback(BiometricAuth.prototype.BIOMETRIC_SUCCESS);
};

module.exports = new BiometricAuth();

require('cordova/exec/proxy').add('BiometricAuth', module.exports);

src/android/BiometricAuth.java
package com.andreszs.biometricauth;

import org.apache.cordova.*;
import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONException;

import android.app.Activity;
import android.app.KeyguardManager;
import android.content.Intent;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.biometric.BiometricPrompt;
import androidx.biometric.BiometricManager;
import androidx.biometric.BiometricManager.Authenticators;
import androidx.core.content.ContextCompat;

import java.util.concurrent.Executor;

interface Constants {
    static final String AUTHENTICATION_FAILED = "AUTHENTICATION_FAILED";
    static final String AUTHENTICATION_SUCCEEDED = "AUTHENTICATION_SUCCEEDED";
    static final String PIN_OR_PATTERN_DISMISSED = "PIN_OR_PATTERN_DISMISSED";
    static final String KEYGUARD_SUCCESS = "KEYGUARD_SUCCESS";
    static final String METHOD_AUTHENTICATE = "authenticate";
    static final String METHOD_IS_AVAILABLE = "isAvailable";

    static final String BIOMETRIC_SUCCESS = "BIOMETRIC_SUCCESS";
    static final String BIOMETRIC_ERROR_UNSUPPORTED = "BIOMETRIC_ERROR_UNSUPPORTED";
    static final String BIOMETRIC_ERROR_NONE_ENROLLED = "BIOMETRIC_ERROR_NONE_ENROLLED";
    static final String BIOMETRIC_ERROR_HW_UNAVAILABLE = "BIOMETRIC_ERROR_HW_UNAVAILABLE";
    static final String BIOMETRIC_ERROR_NO_HARDWARE = "BIOMETRIC_ERROR_NO_HARDWARE";
    static final String BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED = "BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED";
    static final String BIOMETRIC_STATUS_UNKNOWN = "BIOMETRIC_STATUS_UNKNOWN";

    static final int REQUEST_CODE_CONFIRM_DEVICE_CREDENTIALS = 2;
    static final int AUTHENTICATOR_KEYGUARD_MANAGER = 1;
}

public class BiometricAuth extends CordovaPlugin {
    private CallbackContext mCallbackContext;
    private BiometricPrompt mBiometricPrompt;
    private PluginResult pluginResult;
    private static final String TAG = "BiometricAuth";

    @Override
    public boolean execute(String action, JSONArray jsonArgs, CallbackContext callbackContext) throws JSONException {
        mCallbackContext = callbackContext;

        if (action.equals(Constants.METHOD_AUTHENTICATE)) {
            executeAuthenticate(jsonArgs);
            return true;
        } else if (action.equals(Constants.METHOD_IS_AVAILABLE)) {
            executeIsAvailable(jsonArgs);
            return true;
        } else {
            Log.e(TAG, String.format("Invalid action passed: %s", action));
            pluginResult = new PluginResult(PluginResult.Status.INVALID_ACTION);
            callbackContext.sendPluginResult(pluginResult);
            return false;
        }
    }

    private void executeIsAvailable(JSONArray jsonArgs) {
        Log.d(TAG, "executeCanAuthenticate");
        int mAllowedAuthenticators = this.getInt(jsonArgs, "authenticators", 0);

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            /* API <= 20: No API available */
            mCallbackContext.error(Constants.BIOMETRIC_ERROR_UNSUPPORTED);
        } else if (mAllowedAuthenticators == 1 || Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            /* API 21-22: BiometricManager not available, or forced KeyguardManager by params */
            if (mAllowedAuthenticators < 2) {
                /* authenticator 0 = any; 1 = KeyguardManager */
                KeyguardManager keyguardManager = ContextCompat.getSystemService(cordova.getContext(), KeyguardManager.class);
                if (keyguardManager == null) {
                    mCallbackContext.error(Constants.BIOMETRIC_ERROR_UNSUPPORTED);
                } else if (keyguardManager.isKeyguardSecure()) {
                    mCallbackContext.success(Constants.KEYGUARD_SUCCESS);
                } else {
                    mCallbackContext.error(Constants.BIOMETRIC_ERROR_NONE_ENROLLED);
                }
            } else {
                /* not supported on API 21-22 */
                mCallbackContext.error(Constants.BIOMETRIC_ERROR_UNSUPPORTED);
            }
        } else {
            /* API >= 23: Check if the user can authenticate with biometrics. This requires at least one biometric sensor to be present, enrolled, and available on the device. */
            int canAuthenticate;
            int authenticators = getInt(jsonArgs, "authenticators", 0);

            /* Note that not all combinations of authenticator types are supported prior to Android 11 (API 30). Specifically, DEVICE_CREDENTIAL alone is unsupported prior to API 30, and BIOMETRIC_STRONG | DEVICE_CREDENTIAL is unsupported on API 28-29 */
            if (authenticators > 0) {
                /* Return result for specified authenticator(s) */
                canAuthenticate = BiometricManager.from(cordova.getContext()).canAuthenticate(authenticators);
                if (canAuthenticate == BiometricManager.BIOMETRIC_SUCCESS) {
                    mCallbackContext.success(Constants.BIOMETRIC_SUCCESS);
                } else {
                    mCallbackContext.error(getMessageByCode(canAuthenticate));
                }
            } else {
                canAuthenticate = BiometricManager.from(cordova.getContext()).canAuthenticate();
                if (canAuthenticate == BiometricManager.BIOMETRIC_SUCCESS) {
                    /* Any biometric enrolled */
                    mCallbackContext.success(getMessageByCode(canAuthenticate));
                } else {
                    /* Check if KeyguardManager auth is available */
                    KeyguardManager keyguardManager = ContextCompat.getSystemService(cordova.getContext(), KeyguardManager.class);
                    if (keyguardManager != null && keyguardManager.isKeyguardSecure()) {
                        mCallbackContext.success(Constants.KEYGUARD_SUCCESS);
                    } else {
                        mCallbackContext.error(getMessageByCode(canAuthenticate));
                    }
                }
            }

        }
    }

    private void executeAuthenticate(JSONArray jsonArgs) {
        Log.d(TAG, "executeAuthenticate");
        int mAllowedAuthenticators = this.getInt(jsonArgs, "authenticators", 0);

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            mCallbackContext.error(Constants.BIOMETRIC_ERROR_UNSUPPORTED);
        } else if (mAllowedAuthenticators == Constants.AUTHENTICATOR_KEYGUARD_MANAGER || Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            showKeyguardManagerAuth(jsonArgs);
        } else {
            showBiometricManagerAuth(jsonArgs);
        }

    }

    private BiometricPrompt.AuthenticationCallback mAuthenticationCallback = new BiometricPrompt.AuthenticationCallback() {
        @Override
        public void onAuthenticationError(int errorCode, @NonNull CharSequence errString) {
            super.onAuthenticationError(errorCode, errString);
            pluginResult = new PluginResult(PluginResult.Status.ERROR, errString.toString());
            mCallbackContext.sendPluginResult(pluginResult);
        }

        @Override
        public void onAuthenticationSucceeded(@NonNull BiometricPrompt.AuthenticationResult result) {
            super.onAuthenticationSucceeded(result);
            pluginResult = new PluginResult(PluginResult.Status.OK, Constants.AUTHENTICATION_SUCCEEDED);
            mCallbackContext.sendPluginResult(pluginResult);
        }

        @Override
        public void onAuthenticationFailed() {
            super.onAuthenticationFailed();
            pluginResult = new PluginResult(PluginResult.Status.ERROR, Constants.AUTHENTICATION_FAILED);
            mCallbackContext.sendPluginResult(pluginResult);
        }
    };

    private void finishWithError(String message) {
        PluginResult result = new PluginResult(PluginResult.Status.ERROR, message);
        mCallbackContext.sendPluginResult(result);
    }

    private void showBiometricManagerAuth(JSONArray jsonArgs) {
        cordova.getActivity().runOnUiThread(() -> {
            Boolean disableBackup = this.getBoolean(jsonArgs, "disableBackup", false);
            String title = this.getString(jsonArgs, "title", "Enter unlock credentials");
            String subtitle = this.getString(jsonArgs, "subtitle", "");
            String negativeButtonText = this.getString(jsonArgs, "negativeButtonText", "");
            int mAllowedAuthenticators = this.getInt(jsonArgs, "authenticators", 0);

            final Handler handler = new Handler(Looper.getMainLooper());
            Executor executor = handler::post;
            mBiometricPrompt = new BiometricPrompt(cordova.getActivity(), executor, mAuthenticationCallback);
            BiometricPrompt.PromptInfo.Builder promptInfo = new BiometricPrompt.PromptInfo.Builder();
            promptInfo.setTitle(title);
            promptInfo.setSubtitle(subtitle);

            if (mAllowedAuthenticators == 0) {
                if (disableBackup) {
                    mAllowedAuthenticators = Authenticators.BIOMETRIC_WEAK;
                } else {
                    mAllowedAuthenticators = Authenticators.BIOMETRIC_WEAK + Authenticators.DEVICE_CREDENTIAL;
                }
            }
            promptInfo.setAllowedAuthenticators(mAllowedAuthenticators);

            if (!isDeviceCredentialAllowed(mAllowedAuthenticators)) {
                promptInfo.setNegativeButtonText(negativeButtonText);
            }

            try {
                mBiometricPrompt.authenticate(promptInfo.build());
            } catch (Exception e) {
                mCallbackContext.error(e.getMessage());
            }
        });
    }

    private void showKeyguardManagerAuth(JSONArray jsonArgs) {
        String title = this.getString(jsonArgs, "title", "Enter unlock credentials");
        String subtitle = this.getString(jsonArgs, "subtitle", "");

        pluginResult = new PluginResult(PluginResult.Status.NO_RESULT);
        pluginResult.setKeepCallback(true);

        KeyguardManager keyguardManager = ContextCompat.getSystemService(cordova.getContext(), KeyguardManager.class);
        if (keyguardManager == null) {
            mCallbackContext.error(Constants.BIOMETRIC_ERROR_UNSUPPORTED);
        } else if (keyguardManager.isKeyguardSecure()) {
            Intent intent = keyguardManager.createConfirmDeviceCredentialIntent(title, subtitle);
            // Result will be handled by onActivityResult.
            cordova.startActivityForResult((CordovaPlugin) this, intent, Constants.REQUEST_CODE_CONFIRM_DEVICE_CREDENTIALS);
        } else {
            // Show a message that the user hasn't set up a lock screen.
            mCallbackContext.error(Constants.BIOMETRIC_ERROR_NONE_ENROLLED);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == Constants.REQUEST_CODE_CONFIRM_DEVICE_CREDENTIALS) {
            Log.e(TAG, String.valueOf(resultCode));
            if (resultCode == Activity.RESULT_OK) {
                mCallbackContext.success(Constants.AUTHENTICATION_SUCCEEDED);
            } else {
                mCallbackContext.error(Constants.PIN_OR_PATTERN_DISMISSED);
            }
        }
    }

    static Boolean getBoolean(@NonNull JSONArray jsonArray, String key, Boolean defaultValue) {
        try {
            if (!jsonArray.isNull(0)) {
                JSONObject jsonObject = jsonArray.getJSONObject(0);
                if (jsonObject.has(key) && !jsonObject.isNull(key)) {
                    return jsonObject.getBoolean(key);
                }
            }
        } catch (JSONException e) {
            Log.e(TAG, "Can't parse '" + key + "'. Default will be used.", e);
        }
        return defaultValue;
    }

    static int getInt(@NonNull JSONArray jsonArray, String key, int defaultValue) {
        try {
            if (!jsonArray.isNull(0)) {
                JSONObject jsonObject = jsonArray.getJSONObject(0);
                if (jsonObject.has(key) && !jsonObject.isNull(key)) {
                    return jsonObject.getInt(key);
                }
            }
        } catch (JSONException e) {
            Log.e(TAG, "Can't parse '" + key + "'. Default will be used.", e);
        }
        return defaultValue;
    }

    static String getString(@NonNull JSONArray jsonArray, String key, String defaultValue) {
        try {
            if (!jsonArray.isNull(0)) {
                JSONObject jsonObject = jsonArray.getJSONObject(0);
                if (jsonObject.has(key) && !jsonObject.isNull(key)) {
                    return jsonObject.getString(key);
                }
            }
        } catch (JSONException e) {
            Log.e(TAG, "Can't parse '" + key + "'. Default will be used.", e);
        }
        return defaultValue;
    }

    /**
     * Get message string by result code.
     *
     * @param {BiometricManager.Constants} result constants.
     * @return Message code as String.
     */
    static String getMessageByCode(int code) {
        switch (code) {
            case BiometricManager.BIOMETRIC_SUCCESS:
                return Constants.BIOMETRIC_SUCCESS;
            case BiometricManager.BIOMETRIC_ERROR_UNSUPPORTED:
                return Constants.BIOMETRIC_ERROR_UNSUPPORTED;
            case BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE:
                return Constants.BIOMETRIC_ERROR_HW_UNAVAILABLE;
            case BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED:
                return Constants.BIOMETRIC_ERROR_NONE_ENROLLED;
            case BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE:
                return Constants.BIOMETRIC_ERROR_NO_HARDWARE;
            case BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED:
                return Constants.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED;
            default:
                return Constants.BIOMETRIC_STATUS_UNKNOWN;
        }
    }

    /**
     * Checks if a device credential is included in the given set of allowed authenticator types.
     *
     * @param authenticators A bit field representing a set of allowed authenticator types.
     * @return Whether {@link Authenticators#DEVICE_CREDENTIAL} is an allowed authenticator type.
     */
    static boolean isDeviceCredentialAllowed(int authenticators) {
        return (authenticators & Authenticators.DEVICE_CREDENTIAL) != 0;
    }

}
